spec/classes/init_spec.rb

```diff
@@ -968,6 +968,106 @@
+        {
+          :kernel                    => 'Linux',
+          :osfamily                  => 'RedHat',
+          :lsbmajdistrelease         => '6',
+          :operatingsystemmajrelease => '6',
+          :fqdn                      => 'host.example.com',
+          :domain                    => 'example.com',
+          :vas_version               => '4.1.0.21518',
+        }
+      end
+      let :params do
+        {
+          :domain_change  => false,
+          :realm          => 'example.com',
+        }
+      end
+      it { should contain_class('vas') }
+      it { should_not contain_exec('vas_change_domain') }
+    end
+
+    context 'with domain_change set to false and mismatching domains' do
+      let :facts do
+        {
+          :kernel                     => 'Linux',
+          :osfamily                   => 'RedHat',
+          :lsbmajdistrelease          => '6',
+          :operatingsystemmajrelease  => '6',
+          :fqdn                       => 'host.example.com',
+          :domain                     => 'example.com',
+          :vas_domain                 => 'example.com',
+          :vas_version                => '4.1.0.21518',
+        }
+      end
+      let :params do
+        {
+          :domain_change  => false,
+          :realm          => 'example.io',
+        }
+      end
+      it 'should fail' do
+        expect { should contain_class('vas') }.to raise_error(Puppet::Error, /VAS domain missmatch!/)
+      end
+    end
+
+    context 'with domain_change set to true and matching domains' do
+      let :facts do
+        {
+          :kernel                     => 'Linux',
+          :osfamily                   => 'RedHat',
+          :lsbmajdistrelease          => '6',
+          :operatingsystemmajrelease  => '6',
+          :fqdn                       => 'host.example.com',
+          :domain                     => 'example.com',
+          :vas_domain                 => 'example.com',
+          :vas_version                => '4.1.0.21518',
+        }
+      end
+      let :params do
+        {
+          :domain_change  => true,
+          :realm          => 'example.com',
+        }
+      end
+      it { should_not contain_exec('vas_change_domain') }
+    end
+
+    context 'with domain_change set to true and mismatching domains' do
+      let :facts do
+        {
+          :kernel                     => 'Linux',
+          :osfamily                   => 'RedHat',
+          :lsbmajdistrelease          => '6',
+          :operatingsystemmajrelease  => '6',
+          :fqdn                       => 'host.example.com',
+          :domain                     => 'example.com',
+          :vas_domain                 => 'example.com',
+          :vas_version                => '4.1.0.21518',
+        }
+      end
+      let :params do
+        {
+          :domain_change  => true,
+          :realm          => 'example.io',
+        }
+      end
+      it { should contain_exec('vas_change_domain') }
```

Comment:
please also specify the parameters for the exec as they are really essential:
command  => "$(sed 's/\\(.*\\)join.*/\\1unjoin/' /etc/opt/quest/vas/lastjoin) > /tmp/vas_unjoin.txt 2>&1 && rm -f ${once_file}",
onlyif   => "/usr/bin/test -f ${keytab_path} || /usr/bin/test -f /etc/opt/quest/vas/lastjoin",
provider => 'shell',
path     => '/bin:/usr/bin:/opt/quest/bin',
timeout  => 1800,
before   => [File['vas_config'], File['keytab'], Exec['vasinst']],
require  => [Package['vasclnt','vasyp','vasgp']]

You could then remove the other tests for before & require.